Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -50,39 +47,279 @@ extension Workspace {
+        private let sharedMirrors: Mirrors?
+        private let fileSystem: FileSystem
+
+        private var _mirrors: DependencyMirrors
+        private let lock = Lock()
+
+        /// The mirrors in this configuration
+        public var mirrors: DependencyMirrors {
+            self.lock.withLock {
+                self._mirrors
+            }
+        }
 
+        /// A convenience initializer for creating a workspace mirrors configuration for the given root
+        /// package path.
+        ///
+        /// - Parameters:
+        ///   - forRootPackage: The path for the root package.
+        ///   - sharedMirrorFile: Path to the shared mirrors configuration file, defaults to the standard location.
+        ///   - fileSystem: The file system to use.
+        public init(
+            forRootPackage rootPath: AbsolutePath,
+            sharedMirrorFile: AbsolutePath?,
+            fileSystem: FileSystem
+        ) throws {
+            let localMirrorConfigFile = Workspace.DefaultLocations.mirrorsConfigurationFile(forRootPackage: rootPath)
+            try self.init(
+                localMirrorFile: localMirrorConfigFile,
+                sharedMirrorFile: sharedMirrorFile,
+                fileSystem: fileSystem
+            )
+        }
+
+        /// Initialize the workspace mirrors configuration
+        ///
+        /// - Parameters:
+        ///   - localMirrorFile: Path to the workspace mirrors configuration file
+        ///   - sharedMirrorFile: Path to the shared mirrors configuration file, defaults to the standard location.
+        ///   - fileSystem: The file system to use.
+        public init(
+            localMirrorFile: AbsolutePath,
+            sharedMirrorFile: AbsolutePath?,
+            fileSystem: FileSystem
+        ) throws {
+            self.localMirrors = .init(path: localMirrorFile, fileSystem: fileSystem, deleteWhenEmpty: true)
+            self.sharedMirrors = sharedMirrorFile.map { .init(path: $0, fileSystem: fileSystem, deleteWhenEmpty: false) }
+            self.fileSystem = fileSystem
+            // computes the initial mirrors
+            self._mirrors = DependencyMirrors()
+            try self.computeMirrors()
+        }
+
+        @discardableResult
+        public func applyLocal(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            try self.localMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        @discardableResult
+        public func applyShared(handler: (inout DependencyMirrors) throws -> Void) throws -> DependencyMirrors {
+            guard let sharedMirrors = self.sharedMirrors else {
+                throw InternalError("shared mirrors not configured")
+            }
+            try sharedMirrors.apply(handler: handler)
+            try self.computeMirrors()
+            return self.mirrors
+        }
+
+        // 👀 is this correct, or should we merge?
```

Comment:
👀 @neonichu @abertelrud @mattt @yim-lee wdyt?

Reply:
Local mirrors should override shared configuration, and this looks to be doing that. It'd be nice if we could communicate to the user when this happens, so maybe this could be refactored into something like:

```swift
static func resolve(local: Mirrors, shared: Mirrors, diagnosticsEngine: DiagnosticsEngine? = nil) -> Mirrors { /* ... */ }
```